gate-deck-plugins/src/main/kotlin/com/netflix/spinnaker/gate/plugins/DeckPluginCache.kt

```diff
@@ -0,0 +1,145 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.gate.plugins
+
+import com.netflix.spectator.api.Id
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.kork.plugins.bundle.PluginBundleExtractor
+import com.netflix.spinnaker.kork.plugins.update.PluginUpdateService
+import com.netflix.spinnaker.kork.plugins.update.SpinnakerUpdateManager
+import org.pf4j.update.PluginInfo
+import org.slf4j.LoggerFactory
+import org.springframework.scheduling.annotation.Scheduled
+import org.springframework.stereotype.Component
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.StandardCopyOption
+
+/**
+ * Responsible for keeping an up-to-date cache of all plugins that Deck needs to know about.
+ *
+ * TODO(rz): Currently downloads all plugins and inspects the extracted files to determine if a plugin has deck assets.
+ *  This is pretty inefficient - would be better to have a metadata file shipped alongside a bundle.
+ * TODO(rz): Downloads all plugins to an isolated temp directory, so if Gate itself has plugins, this cache will be
+ *  downloading an intersection of plugins twice.
+ */
+@Component
+class DeckPluginCache(
+  private val updateManager: SpinnakerUpdateManager,
+  private val updateService: PluginUpdateService,
+  private val pluginBundleExtractor: PluginBundleExtractor,
+  private val registry: Registry
+) {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  private val cache: MutableSet<PluginCacheEntry> = mutableSetOf()
+  private var cachePopulated: Boolean = false
+
+  private val versionsId = registry.createId("plugins.deckCache.versions")
+  private val hitsId = registry.createId("plugins.deckCache.hits")
+  private val missesId = registry.createId("plugins.deckCache.misses")
+  private val downloadDurationId = registry.createId("plugins.deckCache.downloadDuration")
+  private val refreshDurationId = registry.createId("plugins.deckCache.refreshDuration")
+
+  /**
+   * Refreshes the local file cache of _current_ plugins. Should Deck need plugin assets from an older plugin release
+   * version, it will be downloaded and cached on-demand.
+   *
+   * The default refresh interval is 5 minutes.
+   */
+  @Scheduled(fixedDelayString = "\${spinnaker.extensibility.deck.cache-refresh-interval-ms:300000}", initialDelay = 0)
+  internal fun refresh() {
+    registry.timer(refreshDurationId).record {
+      log.info("Refreshing plugin cache")
+
+      updateManager.refresh()
+
+      val newCache = updateManager.plugins
+        .map {
+          val plugin = DeckPluginVersion(it.id, selectPluginVersion(it))
+          val path = getOrDownload(plugin.id, plugin.version)
```

Comment:
Looks like this will download the specified plugins as per the `DeckPluginVersion`.  I wonder though, if we want to use the `requires` field on Plugin Info `releases`, and search for plugins that require Deck.

Reply:
I was going to do that, but not all plugins will be bundles, and `deck` doesn't have real versions, so any entry for it in `requires` would be totally bogus anyway.

Reply:
OSS deck doesn't really have versions, this is true.  Deck at netflix, however, is composed of published packages, each of which have released versions.  In the future, OSS deck will also be composed of published packages, so we may be able to leverage the `requires` field (eventually).  I would actually encourage any current deck plugins to declare which version of the deck plugin package they're compatible with from the get-go.